community/values/src/main/java/org/neo4j/values/storable/TemporalValue.java

```diff
@@ -1366,6 +1367,32 @@ static int validNano( AnyValue millisecond, AnyValue microsecond, AnyValue nanos
         return (int) (ms * 1000_000 + us * 1000 + ns);
     }
 
+    static <TEMP extends Temporal> TEMP withTruncatedNano( TEMP value )
+    {
+        int precision = Values.getNanoPrecision();
+        if ( precision < 0 || precision > 9 )
+        {
+            throw new InvalidValuesArgumentException( "The dbms.temporal.nanosecond_precision config " +
+                    "must be an integer between 0 and 9, was " + precision );
+        }
+        if ( precision == 9 )
+        {
+            return value;
+        }
+        int truncateFactor = (int) Math.pow( 10, 9 - precision );
```

Comment:
Converting to and from a `double` is actually a reasonably slow way to do this which, at scale, might matter. I've run some quick tests on alternatives and found the most efficient method is simply a hard-coded `switch` list. There are only a handful of allowed values, and they don't change, so this should not make the code hard to maintain.

Compared to the original code, which runs 1m truncations in ~24ms, this completed the same in ~7ms:
```
    private int truncate4(int nanos, int precision)
    {
        switch (precision)
        {
        case 0: return 0;
        case 1: return nanos / 100000000 * 100000000;
        case 2: return nanos / 10000000 * 10000000;
        case 3: return nanos / 1000000 * 1000000;
        case 4: return nanos / 100000 * 100000;
        case 5: return nanos / 10000 * 10000;
        case 6: return nanos / 1000 * 1000;
        case 7: return nanos / 100 * 100;
        case 8: return nanos / 10 * 10;
        case 9: return nanos;
        default:
            throw new IllegalArgumentException("Illegal precision");
        }
    }
```

The other alternatives were (1) multiplying an int by 10 a number of times (~13ms) and (2) looking up in a static `Map` (~19ms).